File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.cpp
class Solution {
 public:
  int minCostClimbingStairs(vector<int>& cost) {
    vector<int> min_sum(cost.size() + 1);
    for (int i = 2; i < min_sum.size(); i++) {
      min_sum[i] =
          min(min_sum[i - 1] + cost[i - 1], min_sum[i - 2] + cost[i - 2]);
    }
    return min_sum[min_sum.size() - 1];
  }
};

File: Python/0746-min-cost-climbing-stairs.py
from functools import lru_cache
from typing import List


class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        @lru_cache
        def minCost(i):
            if i <= 1:
                return 0
            downOne = cost[i-1] + minCost(i-1)
            downTwo = cost[i-2] + minCost(i-2)
            return min(downOne, downTwo)
        return minCost(len(cost))


cost = [10, 15, 20]
print(Solution().minCostClimbingStairs(cost))
